### PR TITLE
Gaming on Solana

### PR DESCRIPTION
**Pull request by blocksan**
Thanks for submitting this tutorial for Figment Learn!

Before you hit "Create Pull Request", make sure you have read the [Tutorial Contribution Guidelines](https://www.notion.so/figmentnetworks/Figment-Learn-Contributor-Guidelines-b08691366e144d33a5636d9276a44ec8) and reviewed the [Sample Tutorial Structure](https://www.notion.so/figmentnetworks/Sample-Tutorial-Structure-667ac2aad36f4f94a3ebaca053180b2d). 

Please confirm the following before submitting your tutorial for review - 

- [x] My tutorial follows the [Tutorial Contribution Guidelines](https://www.notion.so/figmentnetworks/Figment-Learn-Contributor-Guidelines-b08691366e144d33a5636d9276a44ec8).

In particular:

- [x] The title of the Pull Request matches the title of my tutorial.
- [x] I've followed the [tutorial template](https://www.notion.so/figmentnetworks/Sample-Tutorial-Structure-667ac2aad36f4f94a3ebaca053180b2d) structure.
- [x] I've [proofread the tutorial](https://grammarly.com) for spelling and grammar mistakes.
- [x] I've run a linter & prettifier to format my code blocks.
- [x] I've thoroughly explained the code, with inline code comments and text explanations.
- [x] I've run the tutorial to test it and the code works.

**NOTE**: We cannot review your Pull Request if it does not meet the standard outlined above.

### CODE REVIEW

**Comment by yash-sharma1** [comment]
Thanks for submitting the tutorial @blocksan I appreciate all the efforts you have put into writing this tutorial. We're going to start the review and will let you know all the changes required before publishing it.

File: solana/gaming-on-solana/gaming-on-solana.md [hunk 1]
```diff
@@ -0,0 +1,1881 @@
...
+
+export default App;
+```
+
+
+
+
+3. Add the source code for the game we want to integrate with our application
+
+    Check the source code of the game at : https://github.com/solanauniversity/solana-games
```
Comment (yash-sharma1) [inline solana/gaming-on-solana/gaming-on-solana.md:1854]: Can you please provide the detailed steps to add the code and integrate it with our above work? Currently it's quite confuisng for beginner to figure what's next :)

File: solana/gaming-on-solana/gaming-on-solana.md [hunk 2]
```diff
@@ -0,0 +1,1881 @@
...
+
+* Once you complete the above tutorial then you can build any pay per use platform for the end users. 
+* You can provide services against which you can charge some from the users.
+* You can extend the solana games learning and build your gaming platform.
+* You can use the knowledge of NFT creation and build NFT marketplaces like opensea
+
+
+## About The Author
+
+I am a passionate blockchain engineer, avid thinker who is currently thriving to build multiple products in the Solana Ecosystem.
```
Comment (yash-sharma1) [inline solana/gaming-on-solana/gaming-on-solana.md:1870]: You should include your name in this section. You can update it to -

This tutorial was written by [Sandeep Ghosh](https://github.com/blocksan). You can find connect with him on [LinkedIn](https://www.linkedin.com/in/ersandy/) & on [Figment Forum](https://community.figment.io).

@blocksan you can join us on Figment Forum & Discord so we can tag you at both the places while announcing the tutorial.

**Review by yash-sharma1 — CHANGES_REQUESTED** [review]
@blocksan I'm done with the initial review of the tutorial so once you make the requested changes we'll do a once more review of it before the final merge. Also, @pyr0gan can you please give it a review before the final changes? :)

File: solana/gaming-on-solana/gaming-on-solana.md [hunk 3]
```diff
@@ -0,0 +1,1881 @@
...
+     },
+   ],
+ ]);
+```
+
+
+**createMetadata**: In the below util function, we are trying to create the metadata for the mint token which was created earlier. Metadata is going to store the artifacts from the NFT which can be in the form of image/gif/video.
+
+
+```
```
[the remark this answers is not in the document]
Comment (yash-sharma1) [reply solana/gaming-on-solana/gaming-on-solana.md:1515]: I mean like this ⬇️  - 
**```javascript**

**Review by pyr0gan — CHANGES_REQUESTED** [review]
Hi @blocksan,

There are some significant changes required for this tutorial to be published. Please review my comments and if you have any questions do not hesitate to stage me in a comment reply. The main concerns are the code blocks need to be given a syntax highlighting type (this is clearly outlined in the [sample tutorial](https://www.notion.so/Sample-Tutorial-Structure-667ac2aad36f4f94a3ebaca053180b2d)) - There are some flaws in the formatting, especially regarding use of bold text and some HTML tags which need to be cleaned up.

There is also an issue with the amount of code blocks and the related explanation - You will need to either break things down and spend some more time explaining the functions being used, or reduce the amount of code blocks. It is best to create a project repository that users can clone and then only highlight the most important code in the tutorial.

I think one of the largest issues is near the end, line 1852 - You have written a tutorial about games on Solana but there isn't really any game content in the tutorial as it is.

Looking forward to your continued work on this tutorial, I think it has a lot of potential however it still needs some polishing before it's ready to be published 🙏

File: solana/gaming-on-solana/gaming-on-solana.md [hunk 4]
```diff
@@ -0,0 +1,1881 @@
...
+ return {
+   status: false,
+   error: "No transaction found"
+ }
+}
+```
+
+
+
+## How to create the NFT on Solana
```
[the remark this answers is not in the document]
Comment (blocksan) [reply solana/gaming-on-solana/gaming-on-solana.md:248]: In either case, @yash-sharma1 and @pyr0gan to resolve the conflict and make it more understandable, I have updated the tutorial with some explanation.

File: solana/gaming-on-solana/gaming-on-solana.md [hunk 5]
```diff
@@ -0,0 +1,1881 @@
...
+
+    Check the source code of the game at : https://github.com/solanauniversity/solana-games
+
+
+
+## What's Next
+
+
+
+* Once you complete the above tutorial then you can build any pay per use platform for the end users.
```
[the remark this answers is not in the document]
Comment (blocksan) [reply solana/gaming-on-solana/gaming-on-solana.md:1862]: @pyr0gan with all due respect, when I write I use Grammarly and I didn't get any issue with this line. Please suggest the correct grammar for this.

File: solana/gaming-on-solana/gaming-on-solana.md [hunk 6]
```diff
@@ -0,0 +1,1881 @@
...
+};
+```
+
+
+
+## DOM to Image util function
+
+To create NFT which can hold the image as the metadata we need to provide the image in the form of base64 encoded string and from base64 to file object which eventually gets uploaded to arweave.
+
+So the prepare the file object of the image, we wrote our custom function which can convert any DOM element to base64 string and then base64 to File object.
```
[the remark this answers is not in the document]
Comment (blocksan) [reply solana/gaming-on-solana/gaming-on-solana.md:570]: @pyr0gan In the first paragraph I tried to explain how metadata will hold the image and in the second paragraph I tried to explain how the base64 image is getting converted from the util function.
Please suggest the best lines.

File: solana/gaming-on-solana/gaming-on-solana.md [hunk 7]
```diff
@@ -0,0 +1,1881 @@
...
+As such no requirements, only a good code editor will work. I prefer to use VS Code.
+
+
+## Folder structure
+
+**Folder structure expected of the react app to run this tutorial**
+
+
+
+* Assuming app name as "DistributeToken"
```
[the remark this answers is not in the document]
Comment (blocksan) [reply solana/gaming-on-solana/gaming-on-solana.md:59]: The list gets long if we try to expand the folders, hence added the file name with proper formatting.

File: solana/gaming-on-solana/gaming-on-solana.md [hunk 8]
```diff
@@ -0,0 +1,1881 @@
...
+* transaction.serialize() : All the data must be stored on the blockchain. To keep the content format agnostic of the programming language used, the data is serialized before storing. 
+* skipPreflight: It’s a bool data type
+    * true: preflight transaction checks for the available methods before sending the transaction, which involves a very little latency.
+    * false: (default value) : It is turned off by default to save the network bandwidth.
+* preflightCommitment: Every transaction on Solana goes through a process of finalization. The longer the transaction has existed on the blockchain, the less likely it is to get reverted. A transaction is reverted when the blockchain realizes later that the transaction was actually not supposed to be allowed. However, the process of finalization takes some time. Depending on the security required in your code, you can choose between single, confirmed and finalized security levels. Single means that there is atleast one “confirmation” given to the transaction by a validator on Solana.
+
+
+```
```
[the remark this answers is not in the document]
Comment (blocksan) [reply solana/gaming-on-solana/gaming-on-solana.md:1166]: This 'preflightCommitment' is an extra detail that I thought to provide to the learner. I am not using this while making this transaction. To make it easy, I am removing this explanation.

File: solana/gaming-on-solana/gaming-on-solana.md [hunk 9]
```diff
@@ -0,0 +1,1881 @@
...
+     )
+   )[0];
+ }
+```
+
+
+
+## Schema mapping of the required classes
+
+All the classes which we declared on the top need to be converted to struct format with valid data types like u8 which is expected from the on-chain program.
```
[the remark this answers is not in the document]
Comment (blocksan) [reply solana/gaming-on-solana/gaming-on-solana.md:1375]: Done.

**Comment by pyr0gan** [comment]
@blocksan Have you had the opportunity to work on our feedback? We'd like to keep this tutorial moving and get you paid! 😄

**Comment by blocksan** [comment]
Hi Rogan,

Yes, I fixed most of the comments except 5 6 points. I am going to work on
them by today or tomorrow.

Thanks & Regards,
Sandeep Ghosh,
Building www.solgames.fun

File: solana/gaming-on-solana/gaming-on-solana.md [hunk 10]
```diff
@@ -0,0 +1,1881 @@
...
+* transaction.serialize() : All the data must be stored on the blockchain. To keep the content format agnostic of the programming language used, the data is serialized before storing. 
+* skipPreflight: It’s a bool data type
+    * true: preflight transaction checks for the available methods before sending the transaction, which involves a very little latency.
+    * false: (default value) : It is turned off by default to save the network bandwidth.
+* preflightCommitment: Every transaction on Solana goes through a process of finalization. The longer the transaction has existed on the blockchain, the less likely it is to get reverted. A transaction is reverted when the blockchain realizes later that the transaction was actually not supposed to be allowed. However, the process of finalization takes some time. Depending on the security required in your code, you can choose between single, confirmed and finalized security levels. Single means that there is atleast one “confirmation” given to the transaction by a validator on Solana.
```
[the remark this answers is not in the document]
Comment (blocksan) [reply solana/gaming-on-solana/gaming-on-solana.md:1163]: It was not needed, hence removed.

**Comment by blocksan** [comment]
All the comments are fixed, please go through the changes and let me know.

File: solana/gaming-on-solana.md [hunk 11]
```diff
@@ -0,0 +1,1674 @@
...
+* Once you complete the above tutorial then you can build any pay per use platform for the end users. 
+* You can provide services against which you can charge some from the users.
+* You can extend the solana games learning and build your gaming platform.
+* You can use the knowledge of NFT creation and build NFT marketplaces like opensea
+
+
+# About The Author
+
+This tutorial is written by Sandeep Ghosh. You can connect with me on [LinkedIn](https://www.linkedin.com/in/ersandy/) & [Figment Forum](https://community.figment.io/).
+I am a passionate blockchain engineer, an avid thinker who is currently thriving to build multiple products in the Solana ecosystem.
```
Comment (yash-sharma1) [inline solana/gaming-on-solana.md:1663]: All looks good to me. Just a small update in the "About The Author" section. In this section we refer to a third person i:e; but here 2nd & 3rd lines refer to the 1st person ("I am a...") so need to update them.

**Comment by yash-sharma1** [comment]
@pyr0gan also will give it a final review to see if any changes are required or not and after that we can merge it :) 
Thanks, @blocksan for all the efforts you've put into writing this amazing tutorial, appreciate it.

**Comment by blocksan** [comment]
@yash-sharma1 Thanks for your kinds words. I fixed your comment for the 1st person.

**Review by yash-sharma1 — APPROVED** [review]

File: solana/gaming-on-solana.md [hunk 12]
```diff
@@ -1847,35 +1643,32 @@ export default App;
...
 * You can provide services against which you can charge some from the users.
 * You can extend the solana games learning and build your gaming platform.
 * You can use the knowledge of NFT creation and build NFT marketplaces like opensea
 
 
-## About The Author
+# About The Author
 
-I am a passionate blockchain engineer, avid thinker who is currently thriving to build multiple products in the Solana Ecosystem.
+This tutorial is written by Sandeep Ghosh. You can connect with the author on [LinkedIn](https://www.linkedin.com/in/ersandy/) & [Figment Forum](https://community.figment.io/).
```
Comment (pyr0gan) [inline solana/gaming-on-solana.md:1662]: Please link to your Forum user page, not just the Forum main page.

File: solana/gaming-on-solana.md [hunk 13]
```diff
@@ -1847,35 +1643,32 @@ export default App;
...
-2. [https://spl.solana.com/token](https://spl.solana.com/token)
-3. [https://docs.solana.com/terminology](https://docs.solana.com/terminology)
-4. [https://www.metaplex.com/](https://www.metaplex.com/)
-5. https://solgames.fun/
\ No newline at end of file
+1. [Solgames.fun - Superset of this tutorial](https://solgames.fun/) 
+2. [Solana Docs](https://docs.solana.com)
+3. [SPL Token](https://spl.solana.com/token)
+4. [Solana Terminologies](https://docs.solana.com/terminology)
+5. [Metaplex Documenation](https://www.metaplex.com)
```
Comment (pyr0gan) [inline solana/gaming-on-solana.md:1674]: Probably better to link directly to the docs site: <https://docs.metaplex.com/>

Documenation -> Documentation

File: solana/gaming-on-solana.md [hunk 14]
```diff
@@ -1847,35 +1643,32 @@ export default App;
...
-1. [https://docs.solana.com/](https://docs.solana.com/)
-2. [https://spl.solana.com/token](https://spl.solana.com/token)
-3. [https://docs.solana.com/terminology](https://docs.solana.com/terminology)
-4. [https://www.metaplex.com/](https://www.metaplex.com/)
-5. https://solgames.fun/
\ No newline at end of file
+1. [Solgames.fun - Superset of this tutorial](https://solgames.fun/) 
+2. [Solana Docs](https://docs.solana.com)
+3. [SPL Token](https://spl.solana.com/token)
+4. [Solana Terminologies](https://docs.solana.com/terminology)
```
Comment (pyr0gan) [inline solana/gaming-on-solana.md:1673]: Terminologies -> Terminology

File: solana/gaming-on-solana.md [hunk 15]
```diff
@@ -1847,35 +1643,32 @@ export default App;
...
 
-1. [https://docs.solana.com/](https://docs.solana.com/)
-2. [https://spl.solana.com/token](https://spl.solana.com/token)
-3. [https://docs.solana.com/terminology](https://docs.solana.com/terminology)
-4. [https://www.metaplex.com/](https://www.metaplex.com/)
-5. https://solgames.fun/
\ No newline at end of file
+1. [Solgames.fun - Superset of this tutorial](https://solgames.fun/) 
+2. [Solana Docs](https://docs.solana.com)
+3. [SPL Token](https://spl.solana.com/token)
```
Comment (pyr0gan) [inline solana/gaming-on-solana.md:1672]: SPL Token Docs

File: solana/gaming-on-solana.md [hunk 16]
```diff
@@ -185,6 +166,48 @@ const createTransferTransaction = async (ownerPubkey, connection, fromTokenAccou
...
+export const connectOrGetPhantomProvider = (connectToWallet) => {
+   if ("solana" in window) {
+     const provider = window.solana;
+     if(connectToWallet && !window.solana.isConnected){
+           window.solana.connect();
+       }
+       if (provider.isPhantom) {
+           return provider;
+       }
+   }else if(connectToWallet){
```
[the remark this answers is not in the document]
Comment (blocksan) [reply solana/gaming-on-solana.md:203]: Due to project-specific configuration, I have turned my default Prettier off. Do not worry, will fix the spaces based on your comment.

File: solana/gaming-on-solana.md [hunk 17]
```diff
@@ -193,18 +216,21 @@ Once we have these information then we have to sign the transaction using the ab
...
 * This utility function will transfer the token from one user wallet to another user's wallet
 * @param {*} provider : provider of the phantom wallet
 * @param {*} connection : connection to the solana cluster
 * @param {*} tokenToTransfer : tokens to be transferred in lamports
 * @param {*} fromTokenAccountPubkey : sender of the token
 * @param {*} toTokenAccountPubkey : receiver of the token
-* @returns
+* @return {object} 
+//{ status: true, signature: "transaction signature hash"} in case of success
+//{ status: true, signature: "transaction signature hash"} in case of failure
```
[the remark this answers is not in the document]
Comment (blocksan) [reply solana/gaming-on-solana.md:233]: nice catch !!!

File: solana/gaming-on-solana.md [hunk 18]
```diff
@@ -1781,11 +1587,6 @@ function App() {
    /**
     * If the status is true, that means transaction got successful and we can proceed
     */
-```
-
-
-
-```
    history.push('/stack')
```
[the remark this answers is not in the document]
Comment (blocksan) [reply solana/gaming-on-solana.md:1590]: @pyr0gan I totally understand that and I personally faced similar issues in one of the products where I had to forcefully reload the page on-load, but here it works as the history stack is very lean for the current product.

**Comment by blocksan** [comment]
@pyr0gan Thanks for this extensive review, trust me on this that never saw this level of detailed review. I am glad that I actually wrote this tutorial which helped me to curate my content in a more structured way.
Thanks again and please have one more glance at the fixes I pushed.

**Review by pyr0gan — APPROVED** [review]
Thanks for your work on this tutorial, and attention to detail. We appreciate it! 🚀 

Edit:  As I am going over this tutorial and preparing to publish it, it seems that you have not made it into a complete project. As it is, this is a collection of code snippets that are loosely related. I see at the top of the tutorial you mentioned a directory structure, and some of the code blocks reference filenames, but it is not consistently applied throughout the tutorial.

I will attempt to provide the necessary changes before publishing, so that the tutorial makes sense and provides readers with an actual React app to follow along with.

**Comment by pyr0gan** [comment]
@blocksan After reviewing it further, I think there are a few points that could be explained further, which would improve the overall flow of the tutorial:

You listed the dapp features:

> Features
>
> The decentralized application (DApp) will include the following features:
>
> 1. Phantom wallet connection
> 2. SOL transaction to a wallet
> 3. Game play (We will be able to pick any open source game and copy the code)
> 4. NFT creation for the game play high score

- There is nowhere in the tutorial that you actually explain how to connect the gameplay component.
- There is also no explanation of how to extract a high score from a game and turn it into an NFT.

I am in the process of making some edits, but I think it would still be a good idea to fully explain what you set out to explain with this tutorial 😄

**Pull request merged**